demos/sample_azure_iot_adu/sample_azure_iot_adu.c

```diff
@@ -835,17 +863,50 @@ static void prvAzureDemoTask( void * pvParameters )
                                                      sampleazureiotPROCESS_LOOP_TIMEOUT_MS );
             configASSERT( xResult == eAzureIoTSuccess );
 
-            /* TODO: REMOVE !xDidDeviceUpdate for NXP once properly implemented */
             if( xProcessUpdateRequest && !xDidDeviceUpdate )
             {
-                xResult = prvDownloadUpdateImageIntoFlash();
-                configASSERT( xResult == eAzureIoTSuccess );
-
-                xResult = prvEnableImageAndResetDevice();
-                configASSERT( xResult == eAzureIoTSuccess );
-
-                xResult = prvSpoofNewVersion();
-                configASSERT( xResult = eAzureIoTSuccess );
+                if( xAzureIoTAduUpdateRequest.xWorkflow.xAction == eAzureIoTADUActionCancel )
+                {
+                    xResult = AzureIoTADUClient_SendAgentState( &xAzureIoTADUClient,
+                                                                &xAzureIoTHubClient,
+                                                                &xADUDeviceProperties,
+                                                                &xAzureIoTAduUpdateRequest,
+                                                                eAzureIoTADUAgentStateIdle,
+                                                                NULL,
+                                                                ucScratchBuffer,
+                                                                sizeof( ucScratchBuffer ),
+                                                                NULL );
+
+                    xProcessUpdateRequest = false;
```

Comment:
To make sure we are playing safe, should xImage be reset here? 
On line 510 we already check `xImage.ulCurrentOffset < xImage.ulImageFileSize`, so if `ulImageFileSize` is reset to zero, it's an extra layer of safety.

Reply:
https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/88b8c8412f18c1e34544539c2b569a89334bda1d/demos/sample_azure_iot_adu/sample_azure_iot_adu.c#L455

I think this call at the beginning of the `prvDownloadUpdateImageIntoFlash()` would be sufficient? If a cancellation gets invoked in the middle of the download, it kicks out of `prvDownloadUpdateImageIntoFlash()` and then next time around, it would get re-init'd.